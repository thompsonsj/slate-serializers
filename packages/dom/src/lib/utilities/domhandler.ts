import { getAttributeValue } from 'domutils'
import { Element, Text } from 'domhandler'
import serializer from 'dom-serializer'
import { parseStyleCssText } from '.'

/**
 * Generate nested mark elements
 *
 * nestedMarkElements should be recursive, but it works
 * so leaving it for now. Can handle a maximum of 5
 * elements. Really shouldn't be any more than that!
 */

export const nestedMarkElementsString = (els: Element[], text: string) => {
  return serializer(nestedMarkElements(els, new Text(text)))
}

export const nestedMarkElements = (els: Element[], element: Element | Text) => {
  if (els.length === 0) {
    return element
  }
  const el1 = els.pop()
  if (el1) {
    el1.children = [element]
    element = el1
  }
  if (!els || els.length === 0) {
    return element
  }
  const el2 = els.pop()
  if (el2) {
    el2.children = [element]
    element = el2
  }
  if (!els || els.length === 0) {
    return element
  }
  const el3 = els.pop()
  if (el3) {
    el3.children = [element]
    element = el3
  }
  if (!els || els.length === 0) {
    return element
  }
  const el4 = els.pop()
  if (el4) {
    el4.children = [element]
    element = el4
  }
  if (!els || els.length === 0) {
    return element
  }
  const el5 = els.pop()
  if (el5) {
    el5.children = [element]
    element = el5
  }
  if (!els || els.length === 0) {
    return element
  }
  return element
}

/**
 * Extract css value from style attribute
 * @param el domhandler Element
 * @param attribute css attribute in camelCase
 * @returns css value or null
 */
export const extractCssFromStyle = (el: Element, attribute: string): string | null => {
  const cssText = el && getAttributeValue(el, 'style')
  if (cssText) {
    const css = parseStyleCssText(cssText)
    if (css[attribute]) {
      return css[attribute]
    }
  }
  return null
}
