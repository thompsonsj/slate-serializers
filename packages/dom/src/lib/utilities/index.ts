export const hasLineBreak = (str: string) => str.match(/[\r\n]+/) !== null

export const prependSpace = (str: string) => str && ` ${str.trim()}`

const camelize = (str: string) => {
  return str.replace(/(?:^|[-])(\w)/g, (a, c) => {
    c = a.substring(0, 1) === '-' ? c.toUpperCase() : c
    return c ? c : ''
  })
}

export const parseStyleCssText = (value: string): { [key: string]: string } => {
  const output: { [key: string]: string } = {}

  if (!value) {
    return output
  }

  const style = value.split(';')

  for (const s of style) {
    const rule = s.trim()

    if (rule) {
      const ruleParts = rule.split(':')
      const key = camelize(ruleParts[0].trim())
      output[key] = ruleParts[1].trim()
    }
  }

  return output
}

export const styleToString = (style: { [key: string]: string }) => {
  return Object.keys(style).reduce(
    (acc, key) =>
      acc +
      key
        .split(/(?=[A-Z])/)
        .join('-')
        .toLowerCase() +
      ':' +
      style[key] +
      ';',
    '',
  )
}

export const isEmptyObject = (obj: any) =>
  obj && Object.keys(obj).length === 0 && Object.getPrototypeOf(obj) === Object.prototype

export const removeEmpty = (obj: {}): {} => {
  return Object.fromEntries(Object.entries(obj).filter(([_, v]) => v != null))
}

/**
 *
 * @param obj an object of any dimension
 * @param args property list to check
 * @returns undefined or property value
 */
export const getNested = (obj: any, ...args: string[]) => {
  return args.reduce((o, level) => o && o[level], obj)
}

export const encodeBreakingEntities = (str: string) => {
  const swapChar = (charToSwap: string) => {
    // that swaps characters to HTML entities
    switch (charToSwap) {
      case '&':
        return '&amp;'
      case '<':
        return '&lt;'
      case '>':
        return '&gt;'
      default:
        return charToSwap
    }
  }
  str = str.replace(/[&<>]/g, (match) => {
    return swapChar(match)
  })

  return str
}

export const intersection = (o1: { [key: string]: any }, o2: { [key: string]: any }) => {
  return Object.keys(o1)
    .concat(Object.keys(o2))
    .sort()
    .reduce((r: string[], a, i, aa) => {
      if (i && aa[i - 1] === a) {
        r.push(a)
      }
      return r
    }, [])
}
