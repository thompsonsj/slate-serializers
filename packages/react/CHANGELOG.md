# Changelog

## [2.3.0](https://github.com/thompsonsj/slate-serializers/compare/react-v2.2.3...react-v2.3.0) (2025-05-16)


### Features

* **react:** flatten config ([#191](https://github.com/thompsonsj/slate-serializers/issues/191)) ([761c651](https://github.com/thompsonsj/slate-serializers/commit/761c651ad49fda360b416ab561b9aac72ed7aa8d))
