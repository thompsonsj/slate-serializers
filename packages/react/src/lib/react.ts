export function react(): string {
  return 'react';
}
