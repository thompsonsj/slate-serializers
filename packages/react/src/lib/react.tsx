// Serializers
export { SlateToReact } from './serializers'

// Configuration objects
export type { Config as SlateToReactConfig } from './config/types'
export { config as slateToReactConfig } from './config/default'
export { config as payloadSlateToReactConfig } from './config/payload'
export { config as slateDemoSlateToReactConfig } from './config/slateDemo'
