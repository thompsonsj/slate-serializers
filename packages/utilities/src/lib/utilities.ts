export const hasLineBreak = (str: string) => str.match(/[\r\n]+/) !== null

export const prependSpace = (str: string) => str && ` ${str.trim()}`

export const isEmptyObject = (obj: any) =>
  obj && Object.keys(obj).length === 0 && Object.getPrototypeOf(obj) === Object.prototype

export const removeEmpty = (obj: {}): {} => {
  return Object.fromEntries(Object.entries(obj).filter(([_, v]) => v != null))
}

/**
 *
 * @param obj an object of any dimension
 * @param args property list to check
 * @returns undefined or property value
 */
export const getNested = (obj: any, ...args: string[]) => {
  return args.reduce((o, level) => o && o[level], obj)
}
