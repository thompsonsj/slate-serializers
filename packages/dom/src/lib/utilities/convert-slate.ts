import { Document, Element, isTag, Text } from 'domhandler'
import { getName } from 'domutils'
import { encode } from 'html-entities'
import { Text as SlateText } from 'slate'
import { config as defaultConfig } from '../config/default'
import { Config } from '../config/types'
import { nestedMarkElements } from './domhandler'
import { decodeBreakingEntities, encodeBreakingEntities } from '@slate-serializers/utilities'
import { intersection } from '@slate-serializers/utilities'

interface IConvertSlate {
  node: any
  config?: Config
  isLastNodeInDocument?: boolean
  customElementTransforms?: any
  transformText?: (text: Text | Element) => any
  transformElement?: (element: Element) => any
  wrapChildren?: (children: any) => any
}

export const convertSlate = ({
  node,
  config = defaultConfig,
  isLastNodeInDocument = false,
  customElementTransforms,
  /* tslint:disable:no-shadowed-variable */
  transformText = (text) => text,
  /* tslint:disable:no-shadowed-variable */
  transformElement = (element) => element,
  wrapChildren = (children) => new Document(children),
}: IConvertSlate) => {
  if (SlateText.isText(node)) {
    const str =
      config.alwaysEncodeBreakingEntities && !config.encodeEntities ? encodeBreakingEntities(node.text) : node.text

    const strLines = config.convertLineBreakToBr ? str.split('\n') : [str]
    const textChildren: (Element | Text)[] = []

    // convert line breaks to br tags
    strLines.forEach((line, index) => {
      const markElements: Element[] = []
      const markTransformKeys = intersection(config.markTransforms || {}, node)

      markTransformKeys.forEach((key) => {
        if (config.markTransforms?.[key]) {
          const markElement = config.markTransforms[key]({ node, attribs: {} })
          if (markElement) {
            markElements.push(markElement)
          }
        }
      })

      Object.keys(config.markMap).forEach((key) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        if ((node as any)[key]) {
          const elements: Element[] = config.markMap[key].map((tagName) => {
            if (config.markTransforms?.[tagName]) {
              return config.markTransforms[tagName]({ node, attribs: {} })
            }
            return new Element(tagName, {}, [])
          }).filter((element) => element !== undefined) as Element[]
          markElements.push(...elements)
        }
      })
      // clone markElements (it gets modified)
      const markElementsClone = [...markElements]
      const textElement = nestedMarkElements(markElements, new Text(line))

      if (
        config.alwaysEncodeCodeEntities &&
        !config.encodeEntities &&
        isTag(textElement) &&
        getName(textElement) === 'pre'
      ) {
        if (config.alwaysEncodeBreakingEntities) {
          // revert the earlier encoding - encode will re-encode
          line = decodeBreakingEntities(line)
        }
        textChildren.push(nestedMarkElements(markElementsClone, new Text(encode(line))))
      } else {
        textChildren.push(textElement)
      }

      if (index < strLines.length - 1) {
        textChildren.push(transformElement(new Element('br', {})))
      }
    })

    return wrapChildren(textChildren.map((child) =>  transformText(child)))
  }

  const children: any[] = node.children
    ? node.children.map((n: any[]) =>
        convertSlate({
          node: n,
          config,
          customElementTransforms,
          transformText,
          transformElement,
          wrapChildren,
        }),
      )
    : []

  let attribs: { [key: string]: string } = {}

  if (config.elementAttributeTransform) {
    attribs = {
      ...attribs,
      ...config.elementAttributeTransform({ node }),
    }
  }

  let element: Element | null = null

  // more complex transforms
  if (customElementTransforms && customElementTransforms[node.type]) {
    element = customElementTransforms[node.type]({ node, attribs, children })
  } else if (config.elementTransforms[node.type]) {
    element = transformElement(config.elementTransforms[node.type]({ node, attribs, children }) as Element)
  }

  // straightforward node to element mapping
  if (!element && config.elementMap[node.type]) {
    element = transformElement(new Element(config.elementMap[node.type], attribs, children))
  }

  // default tag
  if (!element && config.defaultTag && !node.type) {
    element = transformElement(new Element(config.defaultTag, {}, children))
  }

  if (element) {
    return element
  }

  // add line break between inline nodes
  if (config.convertLineBreakToBr && !isLastNodeInDocument) {
    children.push(transformElement(new Element('br', {})))
  }

  return wrapChildren(children)
}
