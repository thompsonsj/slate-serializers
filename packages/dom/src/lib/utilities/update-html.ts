import { getChildren } from 'domutils'
import { Element } from 'domhandler'
import { HtmlUpdaterFunctionMap } from '../config/types'

export const renameTag = (tagName: string, replacementTagName: string): HtmlUpdaterFunctionMap => ({
  [tagName]: (element: Element) => {
    return new Element(replacementTagName, element.attribs, getChildren(element))
  },
})
