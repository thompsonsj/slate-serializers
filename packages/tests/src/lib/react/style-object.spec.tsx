import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { SlateToReact, slateToReactConfig as defaultReactConfig, SlateToReactConfig } from '@slate-serializers/react'
import { styleObjectFixtures } from './../tests'
import { transformStyleObjectToString } from '@slate-serializers/utilities'

const reactConfig: SlateToReactConfig = {
  ...defaultReactConfig,
  elementAttributeTransform: ({ node }) => {
    const style = transformStyleObjectToString(node.style)
    return style ? { style } : undefined
  }
}

describe('style attribute css transforms with postcss', () => {
  for (const fixture of styleObjectFixtures) {
    it(`${fixture.name}`, () => {
      const tree = render(<SlateToReact node={fixture.slate} config={reactConfig} />)
      expect(tree.container).toMatchSnapshot()
    })
  }
})
