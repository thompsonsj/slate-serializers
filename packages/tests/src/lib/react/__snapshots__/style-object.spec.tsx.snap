// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`style attribute css transforms with postcss mark transform 1`] = `
<div>
  <p>
    <strong>
      Paragraph
    </strong>
  </p>
</div>
`;

exports[`style attribute css transforms with postcss mark transforms on multiple marks 1`] = `
<div>
  <p>
    This is editable 
    <strong>
      rich
    </strong>
     text, 
    <i>
      much
    </i>
     better than a 
    <pre>
      &lt;textarea&gt;
    </pre>
    !
  </p>
</div>
`;

exports[`style attribute css transforms with postcss p tag element transform 1`] = `
<div>
  <p>
    <strong>
      Paragraph
    </strong>
  </p>
</div>
`;
